File: Java/src/main/java/QuestionBank/Solution001_100/Solution19.java
package QuestionBank.Solution001_100;

import tool.ListNode;

/**
 * @author windows
 */
public class Solution19 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null){
            return null;
        }
        ListNode fast = head;
        ListNode cur = head;
        ListNode prev = new ListNode(0, head);
        head = prev;
        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }
        while(fast != null){
            prev = prev.next;
            cur = cur.next;
            fast = fast.next;
        }
        prev.next = cur.next;
        return head;
    }


    public static void main(String[] args) {
        Solution19 solution19 = new Solution19();
    }
}

File: C++/src/Solution1_100/Solution19.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution19.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* removeNthFromEnd(ListNode* head, int n) {
  ListNode* dummy = new ListNode(0, head);
  ListNode* fastPtr = head;
  head = dummy;
  for (int i = 0; i < n; ++i) {
    fastPtr = fastPtr->next;
  }
  while (fastPtr != nullptr){
    fastPtr = fastPtr->next;
    head = head->next;
  }
  head->next = head->next->next;
  return dummy->next;
}